spec.html

```diff
@@ -28651,6 +28653,27 @@
               </tr>
               <tr>
                 <td>
+                  0x0024, 0x003C
+                </td>
+                <td>
+                  `$&lt;`
+                </td>
+                <td>
+                  <emu-alg>
+                    1. If _namedCaptures_ is *undefined*, the replacement text is the String `"$&lt;"`.
+                    1. Otherwise,
+                      1. Scan until the next `>`.
+                      1. If none is found, the replacement text is the String `"$&lt;"`.
+                      1. Otherwise,
+                        1. Let the enclosed substring be _groupName_.
+                        1. Let _capture_ be ? Get(_namedCaptures_, _groupName_).
+                        1. If _capture_ is *undefined*, replace the text through `>` with the empty string.
+                        1. Otherwise, replace the text through this following `>` with ? ToString(_capture_).
```

Comment:
How about:

```
Otherwise, let _replacement_ be ? ToString(_capture_) and replace the text through this following `>` with _replacement_.
```

Reply:
That seems completely unambiguous, LGTM

Reply:
Done

Reply:
These are observably different behaviors. We should make sure there's a test262 test which differentiates them (I don't think there is).

Reply:
Here is a test case (which V8 gets right per spec -- great work @schuay)

```js
d8> let re = /foo(?<bar>baz)bing/
d8> let counter = 0, counter2 = 0
d8> re.exec = (str) => {
  let match = RegExp.prototype.exec.call(re, str);
  let g = match.groups.bar;
  Object.defineProperty(match.groups, "bar", {
    get() {
      counter++;
      return { toString() { counter2++; return g; } }
    }
  });
  return match
}
d8> "foobazbing".replace(re, "$<bar>$<bar>")
"bazbaz"
d8> counter
2
d8> counter2
2
```

Reply:
The ambiguity I see is that if there's N replacements, there should be N get calls - I'm hoping there should be only 1.

Reply:
I don't see any ambiguity. The specification clearly says that there are n get calls (both before and after the rewording).

Can you explain in more detail what you'd like the specification to look like? How should the caching work? Also, what is the motivation for this semantic change?

Reply:
Hmm. If "Otherwise, replace the text through this following `>` with ? ToString(_capture_)." only ever does one replacement in this iteration (even if this algorithm runs for N iterations), then I definitely misunderstood.

Regardless, I think that putting observable operations (like a `ToString` call) in a separate step helps ensure clarity.

Reply:
Both before and after the change, there was one Get and one ToString each time you encountered a particular `$<groupName>`. If the same groupName is encountered twice, there are two Get calls and two ToString calls. What do you think it should do?

I am fine with the editorial change as well.

Reply:
Of course this editorial change does not achieve this; but I'd expect the replacement to be only ToStringed once per `groupName`, throughout the entire replacement.

Reply:
Now that the proposal is at Stage 4 and has multiple browser preview implementations, it seems a little late to make these sorts of suggestions. Generally, we try to reserve Stage 3 and later for implementation-driven changes, having worked out the programmer intuition/language design-related concerns in stages 1-3, unless a big surprise comes up which really demands a response. Do you feel strongly about these semantics?

Reply:
Weighing in from V8: I agree with Dan, it doesn't make sense to me to change this at this point. 

I'm not even sure how the change would be phrased in the spec - would we create a copy of the groups object but storing ToString() of each group? Would we modify the original groups object? Would the ToString() values be calculated lazily or eagerly? And if eagerly, do we really want to penalize the common case in which each group is already a String? Note that calling ToString() on a string is a very cheap operation.

+1 for keeping current semantics.

Reply:
I’m not insisting that be changed; just stating my expectation. It shouldn’t penalize the common case, because in the common case, the ToString will not be observable; it’s only if it’s observable that I’d expect it to only happen once (eagerly or lazily, it doesn’t matter to me).

Certainly it’s fine to keep the existing semantics; but saying “we missed this stage 2 concern and now it’s stage 4 so no change is possible” doesn’t make me feel very confident about the thoroughness of the review process on this proposal.

Reply:
But it does penalize the common case; if done eagerly, we'd need to iterate all keys on the `groups` object, while the replacement string may only reference a subset of them. 

And since I suppose we wouldn't want to modify the given `groups` object, we'd have to allocate a new object somewhere to store the `ToString` normalized group values, even in the lazy case.

I didn't get the feeling that anyone was saying "we missed this stage 2 concern".

Reply:
The comment about it “being a little late” seemed like that.

At any rate, allocating a new unobservanle object somewhere to store a string of Type(replacement) is not String, when in the common case it would be, seems like it would be both a cheap and a rare operation. However my concern is not for performance, which can always be improved upon later; but rather for observability, which can not be reduced later.

Reply:
I don't think there was a lack of review. These semantics were thought out, put out for review for 8 months ago, and now had a thorough review by two implementers. I was aware when writing the spec text that you'd get multiple observable reads. As @schuay raises, it's not clear what the alternative would be, and the downside of the current approach does not seem so high.

RegExps are extremely observable in their behavior in a way which I don't see much of a use case for. I raised this as a concern in late 2015, proposing a more minimal RegExp subclassing mechanism, which would've been much easier to implement and with fewer performance cliffs. The proposal was rejected by the committee. We have a specification with very high observability due to that decision; it will impact many new RegExp features going forward no matter whether we decide to do one or two Get/ToString calls. If we did caching, it would also be observable.

Reply:
Fair point :-)

Reply:
Misunderstanding noted, and reverted to the original spec text.